src/Npgsql/MultiHostConnectorPool.cs

```diff
@@ -0,0 +1,319 @@
+﻿using Npgsql.Util;
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Transactions;
+
+namespace Npgsql
+{
+    sealed class MultiHostConnectorPool : ConnectorSource
+    {
+        readonly ConnectorPool[] _pools;
+
+        public MultiHostConnectorPool(NpgsqlConnectionStringBuilder settings, string connString) : base(settings, connString)
+        {
+            var hosts = settings.Host!.Split(',');
+            _pools = new ConnectorPool[hosts.Length];
+            for (var i = 0; i < hosts.Length; i++)
+            {
+                var host = hosts[i].Trim();
+                var port = settings.Port;
+                var portSeparator = host.IndexOf(':');
+                if (portSeparator != -1)
+                {
+                    port = int.Parse(host.Substring(portSeparator + 1));
+                    host = host.Substring(0, portSeparator);
+                }
+                var poolSettings = settings.Clone();
+                poolSettings.Host = host;
+                poolSettings.Port = port;
+                _pools[i] = new ConnectorPool(poolSettings, poolSettings.ConnectionString, this);
+            }
+        }
+
+        static bool IsPreferred(ClusterState state, TargetSessionAttributes preferredType)
+            => state switch
+            {
+                ClusterState.Offline => false,
+                ClusterState.Unknown => true, // We will check compatibility again after refreshing the cluster state
+                ClusterState.PrimaryReadWrite when preferredType == TargetSessionAttributes.Primary || preferredType == TargetSessionAttributes.PreferPrimary
+                    || preferredType == TargetSessionAttributes.ReadWrite  => true,
+                ClusterState.PrimaryReadOnly when preferredType == TargetSessionAttributes.Primary || preferredType == TargetSessionAttributes.PreferPrimary
+                    || preferredType == TargetSessionAttributes.ReadOnly => true,
+                ClusterState.Standby when preferredType == TargetSessionAttributes.Standby || preferredType == TargetSessionAttributes.PreferStandby
+                    || preferredType == TargetSessionAttributes.ReadOnly => true,
+                _ => preferredType == TargetSessionAttributes.Any
+            };
+
+        static bool IsFallbackOrPreferred(ClusterState state, TargetSessionAttributes preferredType)
+            => state switch
+            {
+                ClusterState.Unknown => true, // We will check compatibility again after refreshing the cluster state
+                ClusterState.PrimaryReadWrite when preferredType == TargetSessionAttributes.PreferPrimary || preferredType == TargetSessionAttributes.PreferStandby => true,
```

Comment:
Am confused... If this is **OrPreferred**, shouldn't we also return true for TargetSessionAttributes.Primary and ReadWrite? Also for the other switch arms.

Reply:
Not really. Because ReadWrite and ReadOnly have no fallback, they're handled together with Any.

Reply:
Am I reading the code wrong? If the ClusterState is PrimaryReadWrite, but the user's preferredType is Primary, it looks like it returns false though it should return true, no? In the same way, if the ClusterState is Standby, the code returns false for preferredType Standby.

Reply:
As discussed offline, this method only gets called when there's a fallback distinct from the preferred (i.e. for Preferred*).